media_cdn/dualtoken.rb

```diff
@@ -0,0 +1,197 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# [START mediacdn_dualtoken_sign_token]
+require "base64"
+require "time"
+require "openssl"
+require "ed25519"
+
+##
+# Media CDN uses URL-safe base64 encoding and strips off the padding at the end.
+# Returns a base64-encoded string compatible with Media CDN.
+#
+def base64_encoder value
+  encoded_str = Base64.urlsafe_encode64(value, padding:false)
+end
+
+##
+# header names
+# Returns an array that aggregates all header names
+#
+def header_names headers
+  header_names = []
+  headers.each do |header|
+    header_names.append header[:name]
+  end
+  header_names.join ","
+end
+
+##
+# header pairs
+# Returns an array that aggregates all header names and values,
+# with each follows name=value format
+#
+def header_pairs headers
+  header_pairs = []
+  headers.each do |header|
+    header_pairs.append "#{header[:name]}=#{header[:value]}"
+  end
+  header_pairs.join ","
+end
+
+##
+# Gets the signed URL suffix string for the Media CDN short token URL requests.
```

Comment:
After looking at other code, it looks like we now use yard for docs.

It should something like this:

```ruby
# Gets the signed URL suffix string for the Media CDN short token URL requests.
# One of (`url_prefix`, `full_path`, `path_globs`) must be included in each input.
#
# @param base64_key [String] a secret key as a base64 encoded string.
# @param signature_algorithm [Symbol] an algorithm as `:SHA1`, `:SHA256`, or `:Ed25519`.
# @param start_time [Time] the start time as a Time object.
# @param expiration_time [Time] the expiration time as a Time object.
#                                       If a value is not specified, the expiration time will be set to 5 mins from now.
# @param url_prefix [String] the URL prefix to sign, including protocol.
#                            For example: http://example.com/path/ for URLs under /path
#                            or http://example.com/path?param=1
# @param full_path [String]  a full path to sign, starting with the first '/'.
#                            For example: /path/to/content.mp4
# @param path_globs [String] a set of ','- or '!'- delimited strings.
#                             For example: /tv/*!/film/* to sign paths starting with /tv/ or /film/ in any URL.
# @param session_id [String]  a unique identifier for the session.
# @param data [String] an arbitrary data payload to include in the token.
# @param headers [Array<Hash<Symbol,String>>] array of header name and value pairs.
#                         For example:
#                            [{name: 'foo', value: 'bar'},
#                             {name: 'baz', value: 'qux'}]
# @param ip_ranges [String] a list of comma-separated IP ranges.
#                              Both IPv4 and IPv6 ranges are acceptable.
#                              For example: '203.0.113.0/24,2001:db8:4a7f:a732/64'
#
# @return [String] The Signed URL appended with the query parameters
#     based on the specified URL prefix and configuration.
#
# @raise [ArgumentError] any of the required arguments are missing.
```